File: flutter_app/test/unit/application/use_cases/delete_todo_use_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:habit_todo_app/core/application/use_cases/delete_todo_use_case.dart';
import 'package:habit_todo_app/core/domain/entities/todo_entity.dart';
import 'package:habit_todo_app/core/domain/repositories/todo_repository.dart';

class MockTodoRepository extends Mock implements TodoRepository {}

void main() {
  group('DeleteTodoUseCase', () {
    late DeleteTodoUseCase useCase;
    late MockTodoRepository mockRepository;

    setUp(() {
      mockRepository = MockTodoRepository();
      useCase = DeleteTodoUseCase(mockRepository);
    });

    test('should delete todo with valid id', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.getById('todo_1')).thenAnswer((_) async => existingTodo);
      when(() => mockRepository.delete('todo_1')).thenAnswer((_) async => Future.value());

      // Act
      await useCase.execute(id: 'todo_1');

      // Assert
      verify(() => mockRepository.getById('todo_1')).called(1);
      verify(() => mockRepository.delete('todo_1')).called(1);
    });

    test('should throw ArgumentError if id is empty', () async {
      // Act & Assert
      expect(
        () => useCase.execute(id: ''),
        throwsA(isA<ArgumentError>()),
      );
      verifyNever(() => mockRepository.getById(any()));
      verifyNever(() => mockRepository.delete(any()));
    });

    test('should throw Exception if todo does not exist', () async {
      // Arrange
      when(() => mockRepository.getById('non_existent')).thenAnswer((_) async => null);

      // Act & Assert
      expect(
        () => useCase.execute(id: 'non_existent'),
        throwsA(isA<Exception>()),
      );
      verify(() => mockRepository.getById('non_existent')).called(1);
      verifyNever(() => mockRepository.delete(any()));
    });

    test('should trim id before processing', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      when(() => mockRepository.getById('todo_1')).thenAnswer((_) async => existingTodo);
      when(() => mockRepository.delete('todo_1')).thenAnswer((_) async => Future.value());

      // Act
      await useCase.execute(id: '  todo_1  ');

      // Assert
      verify(() => mockRepository.getById('todo_1')).called(1);
      verify(() => mockRepository.delete('todo_1')).called(1);
    });
  });
}

File: apps/web/test/unit/application/use_cases/delete_todo_use_case.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { DeleteTodoUseCase } from "../../../../src/core/application/use_cases/delete_todo_use_case";
import type { TodoEntity } from "../../../../src/core/domain/entities/todo_entity";
import type { TodoRepository } from "../../../../src/core/domain/repositories/todo_repository";
import { Priority } from "../../../../src/core/domain/value_objects/priority";

describe("DeleteTodoUseCase", () => {
  let mockRepository: TodoRepository;
  let useCase: DeleteTodoUseCase;

  beforeEach(() => {
    mockRepository = {
      findAll: vi.fn(),
      findById: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
    };
    useCase = new DeleteTodoUseCase(mockRepository);
  });

  it("should delete todo when found", async () => {
    const existing: TodoEntity = {
      id: "test-id",
      title: "Test Todo",
      priority: Priority.MEDIUM,
      completed: false,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    vi.mocked(mockRepository.findById).mockResolvedValue(existing);
    vi.mocked(mockRepository.delete).mockResolvedValue();

    const result = await useCase.execute("test-id");

    expect(result.success).toBe(true);
    expect(mockRepository.delete).toHaveBeenCalledWith("test-id");
  });

  it("should fail when todo not found", async () => {
    vi.mocked(mockRepository.findById).mockResolvedValue(null);

    const result = await useCase.execute("non-existent");

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Todo not found");
    }
  });

  it("should fail when id is empty", async () => {
    const result = await useCase.execute("");

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Todo ID is required");
    }
  });
});



